### PR TITLE
[Cloud Synchronization] Configuration for MQTT base cloud Sync

### LINKED ISSUE #341: Cloud Synchronization for Data stored using Data Storage
A mechanism based on MQTT is required for pushing data stored in core-data DB of Edge Orchestration to the cloud. This would require a cloud endpoint and a configuration for pushing this data to the required cloud on a defined Topic.

Example :

```
Host : **broker.mqttdashboard.com**
Topic : **DataStorage**
```

So, data should be pushed to the hosts DataStorage topic.

Additional configuration such as QOS, Authentication should also be provided.

### PR DESCRIPTION
**Pull request by sun-sharma**
Added configuration that can be un-commented for easy cloud based Data Synchronization

Signed-off-by: Sunchit Sharma <sun.sharma@samsung.com>

# Description

A mechanism based on MQTT is required for pushing data stored in core-data DB of Edge Orchestration to the cloud. This would require a cloud endpoint and a configuration for pushing this data to the required cloud on a defined Topic.

Fixes #341 

## Type of change
- [X] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

### The configuration was set up as follows in the docker-compose.yml file
```
app-service-mqtt:
    image: edgexfoundry/docker-app-service-configurable:1.3.1
    ports:
      - "0.0.0.0:48101:48101"
    container_name: edgex-app-service-configurable-mqtt
    hostname: edgex-app-service-configurable-mqtt
    networks:
      edgex-network:
        aliases:
          - edgex-app-service-configurable-mqtt
    environment:
      EDGEX_SECURITY_SECRET_STORE: "false"
      Registry_Host: edgex-core-consul
      Clients_CoreData_Host: edgex-core-data
      Clients_Data_Host: edgex-core-data # For device Services
      Clients_Notifications_Host: edgex-support-notifications
      Clients_Metadata_Host: edgex-core-metadata
      Clients_Command_Host: edgex-core-command
      Clients_Scheduler_Host: edgex-support-scheduler
      Clients_RulesEngine_Host: edgex-kuiper
      Databases_Primary_Host: edgex-redis
      edgex_profile: mqtt-export
      Service_Host: edgex-app-service-configurable-mqtt
      Service_Port: 48101
      MessageBus_SubscribeHost_Host: edgex-core-data
      Binding_PublishTopic: events
      # Added for MQTT export using app service
      WRITABLE_PIPELINE_FUNCTIONS_MQTTSEND_ADDRESSABLE_ADDRESS: broker.mqttdashboard.com
      WRITABLE_PIPELINE_FUNCTIONS_MQTTSEND_ADDRESSABLE_PORT: 1883
      WRITABLE_PIPELINE_FUNCTIONS_MQTTSEND_ADDRESSABLE_PROTOCOL: tcp
      WRITABLE_PIPELINE_FUNCTIONS_MQTTSEND_ADDRESSABLE_TOPIC: "DataStorage"
      WRITABLE_PIPELINE_FUNCTIONS_MQTTSEND_PARAMETERS_AUTORECONNECT: "true"
      WRITABLE_PIPELINE_FUNCTIONS_MQTTSEND_PARAMETERS_RETAIN: "true"
      WRITABLE_PIPELINE_FUNCTIONS_MQTTSEND_PARAMETERS_PERSISTONERROR: "false"
      # WRITABLE_PIPELINE_FUNCTIONS_MQTTSEND_ADDRESSABLE_PUBLISHER:
      # WRITABLE_PIPELINE_FUNCTIONS_MQTTSEND_ADDRESSABLE_USER:
      # WRITABLE_PIPELINE_FUNCTIONS_MQTTSEND_ADDRESSABLE_PASSWORD:
      # WRITABLE_PIPELINE_FUNCTIONS_MQTTSEND_PARAMETERS_QOS: ["your quality or service"]
      # WRITABLE_PIPELINE_FUNCTIONS_MQTTSEND_PARAMETERS_KEY: [your Key]
      # WRITABLE_PIPELINE_FUNCTIONS_MQTTSEND_PARAMETERS_CERT: [your Certificate]
    depends_on:
      - consul
      # - logging  # uncomment if re-enabled remote logging
      - data
```

### The output was seen on the cloud based MQTT client 

Screenshot from 2021-07-27 12-46-29

Please note that as this was a public free broker QOS was not perfect.

**Test Configuration**:
* Firmware version: Ubuntu 20.04
* Hardware: x86-64
* Edge Orchestration Release: 1.0.0

# Checklist:

- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have made corresponding changes to the documentation
- [X] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [X] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by t25kim** [comment]
Could you explain how to set up the cloud environment?
I would appreciated it if you provide the guide document in the [docs](https://github.com/lf-edge/edge-home-orchestration-go/tree/master/docs) folder.

**Comment by sun-sharma** [comment]
> Could you explain how to set up the cloud environment?
> I would appreciated it if you provide the guide document in the [docs](https://github.com/lf-edge/edge-home-orchestration-go/tree/master/docs) folder.

Thanks for the Review. I have added a doc as required.

**Review by t25kim — COMMENTED** [review]
There is already a [service](https://github.com/lf-edge/edge-home-orchestration-go/blob/da65e9430e44a678e6a8b00d9bb7b5c5a51f066c/deployments/datastorage/docker-compose.yml#L54) that uses the `edgexfoundry/docker-app-service-configurable` image in the docker-compose.yml. Why should we run the `app-service-mqtt` service with a different port? 
Is the `app-service-mqtt` service different from `app-service-rules`?

File: deployments/datastorage/docker-compose.yml [hunk 1]
```diff
@@ -57,6 +57,50 @@ services:
     ports:
     - 0.0.0.0:48100:48100/tcp
     read_only: true
+#  app-service-mqtt:
+#    image: edgexfoundry/docker-app-service-configurable:1.1.0
```
[the remark this answers is not in the document]
Comment (nitu-s-gupta) [reply deployments/datastorage/docker-compose.yml:61]: This was done in accordance to the Hanoi's documentation. https://docs.edgexfoundry.org/1.3/getting-started/quick-start/#exporting-data

File: deployments/datastorage/docker-compose.yml [hunk 2]
```diff
@@ -57,6 +57,50 @@ services:
...
+#    ports:
+#      - "0.0.0.0:48101:48101"
```
[the remark this answers is not in the document]
Comment (nitu-s-gupta) [reply deployments/datastorage/docker-compose.yml:63]: This was done in accordance to the Hanoi's documentation. https://docs.edgexfoundry.org/1.3/getting-started/quick-start/#exporting-data

File: deployments/datastorage/docker-compose.yml [hunk 3]
```diff
@@ -57,6 +57,50 @@ services:
...
+#       WRITABLE_PIPELINE_FUNCTIONS_MQTTSEND_ADDRESSABLE_TOPIC: [Name of receiving topic eg "DataStorage"]
+#       WRITABLE_PIPELINE_FUNCTIONS_MQTTSEND_PARAMETERS_AUTORECONNECT: "true"
+#       WRITABLE_PIPELINE_FUNCTIONS_MQTTSEND_PARAMETERS_RETAIN: "true"
+#       WRITABLE_PIPELINE_FUNCTIONS_MQTTSEND_PARAMETERS_PERSISTONERROR: "false"
+#       WRITABLE_PIPELINE_FUNCTIONS_MQTTSEND_ADDRESSABLE_PUBLISHER: [Addressable Publisher]
+#       WRITABLE_PIPELINE_FUNCTIONS_MQTTSEND_ADDRESSABLE_USER: [username]
+#       WRITABLE_PIPELINE_FUNCTIONS_MQTTSEND_ADDRESSABLE_PASSWORD: [password]
+#       WRITABLE_PIPELINE_FUNCTIONS_MQTTSEND_PARAMETERS_QOS: ["your quality or service"]
+#       WRITABLE_PIPELINE_FUNCTIONS_MQTTSEND_PARAMETERS_KEY: [your Key]
+#       WRITABLE_PIPELINE_FUNCTIONS_MQTTSEND_PARAMETERS_CERT: [your Certificate]
```
[the remark this answers is not in the document]
Comment (sun-sharma) [reply deployments/datastorage/docker-compose.yml]: Done

File: docs/cloud_setup_mqtt.md [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+5. [Optional - Testing Setup on A Public Broker - HiveMQ](#5-how-to-setup)
+6. [Links for setup on other vendors](#6-Links-for-setup-on-other-vendors)
+
+## 1. Introduction
+This module would be responsible for sending the data (can be sensor data or any reading or image/video data, etc.) collected from different devices in the home environment to a cloud Endpoint. This Sync to the cloud will be in every few seconds. The user can define the endpoints and provide the configuration for automatic sync to the cloud.
+
+## 2. Local Setup Description
+Data export to an external source, like an MQTT topic, AWS or Azure, can be done by configuring the options in the
+[docker-compose.yml](https://github.com/lf-edge/edge-home-orchestration-go/blob/da65e9430e44a678e6a8b00d9bb7b5c5a51f066c/deployments/datastorage/docker-compose.yml) in deployments/datastorage.
+
```
[the remark this answers is not in the document]
Comment (sun-sharma) [reply docs/cloud_setup_mqtt.md:16]: Added at last of point 3

File: deployments/datastorage/docker-compose.yml [hunk 5]
```diff
@@ -57,6 +57,50 @@ services:
...
+#      Clients_Metadata_Host: edgex-core-metadata
+#      Clients_Command_Host: edgex-core-command
+#      Clients_Scheduler_Host: edgex-support-scheduler
+#      Clients_RulesEngine_Host: edgex-kuiper
+#      Databases_Primary_Host: edgex-redis
+#      edgex_profile: mqtt-export
+#      Service_Host: edgex-app-service-configurable-mqtt
+#      Service_Port: 48101
+#      MessageBus_SubscribeHost_Host: edgex-core-data
+#      Binding_PublishTopic: events
```
[the remark this answers is not in the document]
Comment (sun-sharma) [reply deployments/datastorage/docker-compose.yml]: Done

File: docs/cloud_setup_mqtt.md [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+      Clients_Metadata_Host: edgex-core-metadata
+      Clients_Command_Host: edgex-core-command
+      Clients_Scheduler_Host: edgex-support-scheduler
+      Clients_RulesEngine_Host: edgex-kuiper
+      Databases_Primary_Host: edgex-redis
+      edgex_profile: mqtt-export
+      Service_Host: edgex-app-service-configurable-mqtt
+      Service_Port: 48101
+      MessageBus_SubscribeHost_Host: edgex-core-data
+      Binding_PublishTopic: events
```
[the remark this answers is not in the document]
Comment (sun-sharma) [reply docs/cloud_setup_mqtt.md]: Done

File: deployments/datastorage/docker-compose.yml (hunk 1 above)
[the remark this answers is not in the document]
Comment (t25kim) [reply deployments/datastorage/docker-compose.yml:61]: @nitu-s-gupta Thanks for the detailed information. However, it seems that the guide was written in [Geneva](https://docs.edgexfoundry.org/1.2/getting-started/quick-start/#exporting-data) and has not been updated. Hanoi uses docker-app-service-configurable v1.3.1, so please try to use v1.3.1 if possible.

[the remark this answers is not in the document]
Comment (sun-sharma) [reply deployments/datastorage/docker-compose.yml]: Thanks @t25kim and @nitu-s-gupta, I have updated to version 1.3.1 post testing.

File: docs/cloud_setup_mqtt.md [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+The CloudMQTT dashboard allows you to monitor your current connection and message traffic statistics.
+
+The dashboard can be accessed via the CLI:
+```
+heroku addons:open cloudmqtt
+Opening cloudmqtt for sharp-mountain-4005...
+```
+or by visiting the [Heroku Dashboard](https://dashboard.heroku.com/) and selecting the application in question. Select CloudMQTT from the Add-ons menu.
+
+# 5. Optional - Testing Setup on A Public Broker - HiveMQ
```
Comment (t25kim) [inline docs/cloud_setup_mqtt.md:143]: This is a very useful guide, but the contents (including pictures) seem to be derived from edgex. 
If right, how about guiding this to its original link? You can put short comments with the link on this document.

Comment (sun-sharma) [reply docs/cloud_setup_mqtt.md:143]: @t25kim Some parts (Like point 5) have been derived from EdgeX. Other parts have been written by me. Some images have also been drawn by me, while others as in point 5 have been derived. Would you like that I add the link to references (Same as point 3), or remove the data and add the link only ? I personally feel It is better to leave the content for easy readability and continuity.

Comment (t25kim) [reply docs/cloud_setup_mqtt.md:143]: I agree that it is very helpful in terms of readability and continuity. 
What about modifying `HiveMQTT Messages` and `Subscriptions` pictures in accordance with Home Edge?
I think they should be revised to fit our project. If the modification is difficult, leave the link along with brief descriptions.

Comment (sun-sharma) [reply docs/cloud_setup_mqtt.md:143]: Done.

**Review by t25kim — APPROVED** [review]
LGTM Thanks!

**Review by MoonkiHong — APPROVED** [review]
LGTM.

**Review by suresh-lc — APPROVED** [review]
Thanks for the synch addition, LGTM

**Pull request merged**